File: backend/app/models/product_model.py
from typing import Optional

from pydantic import BaseModel, Field


class ProductPublic(BaseModel):
    id: int
    nombre: str
    categoria: str
    precio: float
    imagen: str
    stock: int
    activo: bool


class ProductAdmin(ProductPublic):
    precio_compra: Optional[float] = None
    proveedor: Optional[str] = None
    stock_minimo: Optional[int] = 0


class ProductCreate(BaseModel):
    nombre: str = Field(..., min_length=1)
    categoria: str = Field(..., min_length=1)
    precio: float = Field(..., ge=0)
    imagen: str = ""
    stock: int = Field(default=0, ge=0)
    activo: bool = True
    precio_compra: Optional[float] = Field(default=None, ge=0)
    proveedor: Optional[str] = None
    stock_minimo: Optional[int] = Field(default=0, ge=0)


class ProductUpdate(BaseModel):
    nombre: Optional[str] = None
    categoria: Optional[str] = None
    precio: Optional[float] = Field(default=None, ge=0)
    imagen: Optional[str] = None
    stock: Optional[int] = Field(default=None, ge=0)
    activo: Optional[bool] = None
    precio_compra: Optional[float] = Field(default=None, ge=0)
    proveedor: Optional[str] = None
    stock_minimo: Optional[int] = Field(default=None, ge=0)


class ProductStockUpdate(BaseModel):
    stock: int = Field(..., ge=0)


class ApiMessage(BaseModel):
    message: str
    product_id: Optional[int] = None

File: cash-control-app/lib/modules/business/models/product_model.dart
class ProductModel {
  final int id;
  final String nombre;
  final String categoria;
  final double precio;
  final String imagen;
  final int stock;
  final bool activo;
  final double? precioCompra;
  final String? proveedor;
  final int? stockMinimo;

  ProductModel({
    required this.id,
    required this.nombre,
    required this.categoria,
    required this.precio,
    required this.imagen,
    required this.stock,
    required this.activo,
    this.precioCompra,
    this.proveedor,
    this.stockMinimo,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) {
    return ProductModel(
      id: _toInt(json['id']),
      nombre: json['nombre']?.toString() ?? '',
      categoria: json['categoria']?.toString() ?? '',
      precio: _toDouble(json['precio']),
      imagen: json['imagen']?.toString() ?? '',
      stock: _toInt(json['stock']),
      activo: json['activo'] == true,
      precioCompra: _toNullableDouble(json['precio_compra']),
      proveedor: json['proveedor']?.toString(),
      stockMinimo: _toNullableInt(json['stock_minimo']),
    );
  }

  static int _toInt(dynamic value) {
    if (value == null) return 0;
    if (value is int) return value;
    if (value is double) return value.toInt();
    return int.tryParse(value.toString()) ?? 0;
  }

  static int? _toNullableInt(dynamic value) {
    if (value == null) return null;
    if (value is int) return value;
    if (value is double) return value.toInt();
    return int.tryParse(value.toString());
  }

  static double _toDouble(dynamic value) {
    if (value == null) return 0;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    return double.tryParse(value.toString()) ?? 0;
  }

  static double? _toNullableDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    return double.tryParse(value.toString());
  }
}
